### ISSUE TITLE
[C++][Parquet] Use Arrow IO interfaces natively rather than current parquet:: wrappers

### ISSUE THREAD

**Issue by asfimport** [apache/arrow#42358]
We are beginning to do some work on asynchronous IO in Arrow and it would be great to be able to leverage this in the Parquet core internals. 

I am proposing to remove the Parquet-specific virtual file interfaces in

https://github.com/apache/arrow/blob/master/cpp/src/parquet/util/memory.h#L221

and instead rely directly on the Arrow ones in arrow::io. In addition to reducing the amount of code we have to maintain, we will also be able to improve performance of Parquet by utilizing common utilities for managing asynchronous / background IO

cc @majetideepak @xhochy

**Reporter**: [Wes McKinney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wesm) / @wesm
**Assignee**: Wes McKinney](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wesm) / @wesm
#### Related issues:
- [[C++] Consolidate IO interfaces used in arrow/io and parquet-cpp](https://issues.apache.org/jira/browse/ARROW-3166) (is related to)
#### PRs and other links:
- [GitHub Pull Request #4404](https://github.com/apache/arrow/pull/4404)

<sub>**Note**: *This issue was originally created as [PARQUET-1422](https://issues.apache.org/jira/browse/PARQUET-1422). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Uwe Korn](https://issues.apache.org/jira/browse/PARQUET-1422?#comment-16624704) / @xhochy:
+1 from me, that's one of the refactoring benefits I was hoping for to see.

**Comment by asfimport** [comment]
[James Clampffer](https://issues.apache.org/jira/browse/PARQUET-1422?#comment-16625752):
How about leaving pure virtual interfaces in the parquet headers and have those use the Arrow implementations by default?  It looks like doing this as described will require more Arrow headers to be included in the build of the core parquet library.  This way parquet-core can be slimmed down without adding new dependencies on Arrow.

**Comment by asfimport** [comment]
[Wes McKinney](https://issues.apache.org/jira/browse/PARQUET-1422?#comment-16625840) / @wesm:
hi [~James C] – I am not sure how we can maintain separate virtual interfaces while taking advantage of additional IO facilities (like asynchronous buffering) that require using shared abstractions. This is a proposal to perform a net code deletion.

**Comment by asfimport** [comment]
[Wes McKinney](https://issues.apache.org/jira/browse/PARQUET-1422?#comment-16625848) / @wesm:
Also on 

> doing this as described will require more Arrow headers to be included in the build of the core parquet library

This does not cause Parquet core to depend on additional Arrow headers. This work is in line with having less "platform code" in the Parquet core libraries, one of our objectives in bringing the communities closer together.

I will post a patch sometime soon for comment

**Comment by asfimport** [comment]
[Wes McKinney](https://issues.apache.org/jira/browse/PARQUET-1422?#comment-16850097) / @wesm:
@pitrou I ran into a snag working on this in the different semantics of `parquet::BufferedInputStream` versus `arrow::io::BufferedInputStream`. The only place this function is used in parquet-cpp is here

https://github.com/apache/arrow/blob/1a0e976/cpp/src/parquet/column_reader.cc#L156

The idea is that we don't yet know how big the next page header is so we keep trying to deserialize a larger and larger page header with Thrift until we reach the maximum allowable page size.

`parquet::BufferedInputStream` when passed larger and larger peek sizes will expand the size of the buffer to suit the request. 

The easiest thing would be to port the logic from parquet-cpp to in

https://github.com/apache/arrow/blob/1a0e976/cpp/src/arrow/io/buffered.cc#L258

Another issue I found is that `arrow::io::BufferedInputStream::Peek` cannot trigger buffering unlike the `parquet::` counterpart
